diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/boot/utils/mktree.c linux-2.6.12/arch/ppc/boot/utils/mktree.c
--- linux-2.6.12.orig/arch/ppc/boot/utils/mktree.c	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/boot/utils/mktree.c	2005-06-24 21:41:46.000000000 +0200
@@ -113,7 +113,8 @@
 		exit(4);
 	}

-	nblks -= (64 * 1024) / IMGBLK;
+	nblks -= (64 * 1024 - sizeof (bt)) / IMGBLK;	// missed some data
+	bt.bb_num_512blocks = htonl(nblks + 1);			// better safe than sorry...

 	/* And away we go......
 	*/
@@ -122,7 +123,7 @@
 		exit(5);
 	}

-	while (nblks-- > 0) {
+	while (--nblks > 0) {
 		if (read(in_fd, tmpbuf, IMGBLK) < 0) {
 			perror("zImage read");
 			exit(5);
