diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/boot/simple/embed_config.c linux-2.6.12/arch/ppc/boot/simple/embed_config.c
--- linux-2.6.12.orig/arch/ppc/boot/simple/embed_config.c	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/boot/simple/embed_config.c	2005-06-27 22:56:02.000000000 +0200
@@ -781,14 +781,19 @@
 }
 #endif /* CONFIG_XILINX_ML300 */

-#ifdef CONFIG_IBM_OPENBIOS
+#if defined(CONFIG_IBM_OPENBIOS)
 /* This could possibly work for all treeboot roms.
 */
+// but not for me - sorry .... (I do not want to waste valuable flash from 0030 to 0b50)
+	#if defined(CONFIG_TRIPLEDRAGON) || defined(CONFIG_REDWOOD_5)
+		#define BOARD_INFO_VECTOR	0xFFFE0030
+	#else
 #if defined(CONFIG_ASH) || defined(CONFIG_BEECH) || defined(CONFIG_BUBINGA)
 #define BOARD_INFO_VECTOR       0xFFF80B50 /* openbios 1.19 moved this vector down  - armin */
 #else
 #define BOARD_INFO_VECTOR	0xFFFE0B50
 #endif
+	#endif

 #ifdef CONFIG_BEECH
 static void
@@ -814,8 +819,8 @@
 	bd_t	*bd, *treeboot_bd;
 	bd_t *(*get_board_info)(void) =
 	    (bd_t *(*)(void))(*(unsigned long *)BOARD_INFO_VECTOR);
-#if !defined(CONFIG_STB03xxx)

+#if !defined(CONFIG_STB03xxx) && !defined(CONFIG_TRIPLEDRAGON)
 	/* shut down the Ethernet controller that the boot rom
 	 * sometimes leaves running.
 	 */
@@ -855,7 +860,7 @@
 	 * different object in the structure.  Sincr Redwwood 5
 	 * and Redwood 6 use OpenBIOS, it requires a special value.
 	 */
-#if defined(CONFIG_REDWOOD_5) || defined (CONFIG_REDWOOD_6)
+#if defined(CONFIG_IBM_PALLAS) || defined(CONFIG_REDWOOD_6)
 	bd->bi_tbfreq = 27 * 1000 * 1000;
 #endif
 	timebase_period_ns = 1000000000 / bd->bi_tbfreq;
diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/boot/simple/head.S linux-2.6.12/arch/ppc/boot/simple/head.S
--- linux-2.6.12.orig/arch/ppc/boot/simple/head.S	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/boot/simple/head.S	2005-06-27 23:04:16.000000000 +0200
@@ -35,7 +35,7 @@
 	.globl	start
 start:
 	bl	start_
-#ifdef CONFIG_IBM_OPENBIOS
+#if defined(CONFIG_IBM_OPENBIOS)
 	/* The IBM "Tree" bootrom knows that the address of the bootrom
 	 * read only structure is 4 bytes after _start.
 	 */
diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/boot/simple/misc-embedded.c linux-2.6.12/arch/ppc/boot/simple/misc-embedded.c
--- linux-2.6.12.orig/arch/ppc/boot/simple/misc-embedded.c	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/boot/simple/misc-embedded.c	2005-06-27 23:05:19.000000000 +0200
@@ -82,17 +82,26 @@
 unsigned long
 load_kernel(unsigned long load_addr, int num_words, unsigned long cksum, bd_t *bp)
 {
-	char *cp, ch;
-	int timer = 0, zimage_size;
+#if !defined(CONFIG_TRIPLEDRAGON)
+	char ch;
+	int timer = 0;
+#endif
+	char *cp;
+	int zimage_size;
 	unsigned long initrd_size;

 	/* First, capture the embedded board information.  Then
 	 * initialize the serial console port.
 	 */
 	embed_config(&bp);
+
+#if defined(CONFIG_TRIPLEDRAGON)
+	com_port = serial_init(1, bp);
+#else
 #if defined(CONFIG_SERIAL_CPM_CONSOLE) || defined(CONFIG_SERIAL_8250_CONSOLE)
 	com_port = serial_init(0, bp);
 #endif
+#endif

 	/* Grab some space for the command line and board info.  Since
 	 * we no longer use the ELF header, but it was loaded, grab
@@ -177,6 +186,15 @@
 	 * If we weren't, see if we have a ramdisk.  If so, thats root.
 	 * When in doubt, give them the netroot (root=/dev/nfs rw) -- Tom
 	 */
+
+// thats much better for us. We simply pass a pointer to a comandline inside the "board-info-struct"
+// this cmdline is customizeable within our bootloader
+// we need it for example to customize mtd partitions or nfs mounts (especially for nft root)
+#if defined(CONFIG_TRIPLEDRAGON)
+	if( bp->bi_cmdline )
+		memcpy(cmd_line, bp->bi_cmdline, strlen(bp->bi_cmdline) + 1);
+#else
+
 #ifdef CONFIG_CMDLINE_BOOL
 	memcpy (cmd_line, compiled_string, sizeof(compiled_string));
 #else
@@ -185,8 +203,14 @@
 	else
 		memcpy (cmd_line, netroot_string, sizeof(netroot_string));
 #endif
+
+#endif
+
 	while ( *cp )
 		putc(*cp++);
+
+// time is money
+#if !defined(CONFIG_TRIPLEDRAGON)
 	while (timer++ < 5*1000) {
 		if (tstc()) {
 			while ((ch = getc()) != '\n' && ch != '\r') {
@@ -210,7 +234,10 @@
 		}
 		udelay(1000);  /* 1 msec */
 	}
+#endif
+
 	*cp = 0;
+
 	puts("\nUncompressing Linux...");

 	gunzip(0, 0x400000, zimage_start, &zimage_size);
@@ -266,7 +293,7 @@
 		rec->size = sizeof(struct bi_record);
 		rec = (struct bi_record *)((unsigned long)rec + rec->size);
 	}
-	puts("Now booting the kernel\n");
+	puts("Now booting the kernel (misc-embedded)\n");
 #if defined(CONFIG_SERIAL_CPM_CONSOLE) || defined(CONFIG_SERIAL_8250_CONSOLE)
 	serial_close(com_port);
 #endif
diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/boot/simple/misc.c linux-2.6.12/arch/ppc/boot/simple/misc.c
--- linux-2.6.12.orig/arch/ppc/boot/simple/misc.c	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/boot/simple/misc.c	2005-06-27 13:17:39.000000000 +0200
@@ -99,9 +99,13 @@
 	struct bi_record *rec;
 	unsigned long initrd_loc = 0, TotalMemory = 0;

+#if defined(CONFIG_TRIPLEDRAGON)
+	com_port = serial_init(1, NULL);
+#else
 #if defined(CONFIG_SERIAL_8250_CONSOLE) || defined(CONFIG_SERIAL_MPSC_CONSOLE)
 	com_port = serial_init(0, NULL);
 #endif
+#endif

 #if defined(CONFIG_44x) && defined(PPC44x_EMAC0_MR0)
 	/* Reset MAL */
@@ -263,7 +267,7 @@

 		bootinfo_append(BI_INITRD, sizeof(initrd), &initrd);
 	}
-	puts("Now booting the kernel\n");
+	puts("Now booting the kernel (misc)\n");
 	serial_close(com_port);

 	return rec;
diff -Xxfiles -rNBwdu linux-2.6.12.orig/include/asm-ppc/ibm4xx.h linux-2.6.12/include/asm-ppc/ibm4xx.h
--- linux-2.6.12.orig/include/asm-ppc/ibm4xx.h	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/include/asm-ppc/ibm4xx.h	2005-06-27 22:11:26.000000000 +0200
@@ -43,7 +43,7 @@
 #include <platforms/4xx/redwood.h>
 #endif

-#if defined(CONFIG_REDWOOD_5)
+#if defined(CONFIG_IBM_PALLAS)
 #include <platforms/4xx/redwood5.h>
 #endif

diff -Xxfiles -rNBwdu linux-2.6.12.orig/include/asm-ppc/ppc4xx_dma.h linux-2.6.12/include/asm-ppc/ppc4xx_dma.h
--- linux-2.6.12.orig/include/asm-ppc/ppc4xx_dma.h	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/include/asm-ppc/ppc4xx_dma.h	2005-06-25 21:31:57.000000000 +0200
@@ -182,7 +182,7 @@

 typedef uint32_t sgl_handle_t;

-#ifdef CONFIG_PPC4xx_EDMA
+#if defined(CONFIG_PPC4xx_DMA) && !defined(CONFIG_STBXXX_DMA)

 #define SGL_LIST_SIZE 4096
 #define DMA_PPC4xx_SIZE SGL_LIST_SIZE
@@ -285,6 +285,12 @@

 #define GET_DMA_POLARITY(chan) (DMAReq_ActiveLow(chan) | DMAAck_ActiveLow(chan) | EOT_ActiveLow(chan))

+
+#define DMA_CH_ENABLE (1<<7)
+#define GET_DMA_CH(x) (((x)&DMA_CH_ENABLE)>>7)
+
+
+//--------------------------------------------------------------------------
 #elif defined(CONFIG_STBXXX_DMA)		/* stb03xxx */

 #define DMA_PPC4xx_SIZE	4096
@@ -484,7 +490,7 @@
 	unsigned int mode;	/* transfer mode                     */
 	phys_addr_t addr;
 	char ce;		/* channel enable */
-#ifdef CONFIG_STB03xxx
+#if defined(CONFIG_STB03xxx) || defined(CONFIG_STBXXX_DMA)
 	char ch_enable;
 	char tcd_disable;
 	char ece_enable;
diff -Xxfiles -rNBwdu linux-2.6.12.orig/include/linux/soundcard.h linux-2.6.12/include/linux/soundcard.h
--- linux-2.6.12.orig/include/linux/soundcard.h	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/include/linux/soundcard.h	2005-06-25 15:45:08.000000000 +0200
@@ -574,6 +574,17 @@
 #	define AFMT_MPEG		0x00000200	/* MPEG (2) audio */
 #	define AFMT_AC3		0x00000400	/* Dolby Digital AC3 */

+/* because they *do* exist and we want to use them --Monty */
+#define AFMT_S24_LE		0x00000800
+#define AFMT_S24_BE		0x00001000
+#define AFMT_U24_LE		0x00002000
+#define AFMT_U24_BE		0x00004000
+#define AFMT_S32_LE		0x00008000
+#define AFMT_S32_BE		0x00010000
+#define AFMT_U32_LE		0x00020000
+#define AFMT_U32_BE		0x00040000
+
+
 /*
  * Buffer status queries.
  */
diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/platforms/4xx/Kconfig linux-2.6.12/arch/ppc/platforms/4xx/Kconfig
--- linux-2.6.12.orig/arch/ppc/platforms/4xx/Kconfig	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/platforms/4xx/Kconfig	2005-06-27 22:01:12.000000000 +0200
@@ -61,6 +61,11 @@
 	help
 	  This option enables support for the Xilinx ML300 evaluation board.

+config TRIPLEDRAGON
+	bool "Triple-Dragon"
+	help
+	  This option enables support for the TripleDragon STB.
+
 endchoice

 choice
@@ -142,7 +147,7 @@

 config IBM_OCP
 	bool
-	depends on ASH || BUBINGA || CPCI405 || EBONY || EP405 || LUAN || OCOTEA || REDWOOD_5 || REDWOOD_6 || SYCAMORE || WALNUT
+	depends on ASH || BUBINGA || CPCI405 || EBONY || EP405 || LUAN || OCOTEA || REDWOOD_5 || REDWOOD_6 || SYCAMORE || WALNUT || TRIPLEDRAGON
 	default y

 config XILINX_OCP
@@ -187,7 +192,12 @@

 config STB03xxx
 	bool
-	depends on REDWOOD_5 || REDWOOD_6
+	depends on REDWOOD_5 || REDWOOD_6 || TRIPLEDRAGON
+	default y
+
+config IBM_PALLAS
+	bool
+	depends on REDWOOD_5 || TRIPLEDRAGON
 	default y

 config EMBEDDEDBOOT
@@ -197,7 +207,7 @@

 config IBM_OPENBIOS
 	bool
-	depends on ASH || BUBINGA || REDWOOD_5 || REDWOOD_6 || SYCAMORE || WALNUT
+	depends on ASH || BUBINGA || REDWOOD_5 || REDWOOD_6 || SYCAMORE || WALNUT || TRIPLEDRAGON
 	default y

 config PPC4xx_DMA
@@ -205,10 +215,15 @@
 	depends on 4xx

 config PPC4xx_EDMA
-	bool
+	bool "PPC4xx EDMA (SG) Controller Support"
 	depends on !STB03xxx && PPC4xx_DMA
 	default y

+config STBXXX_DMA
+	bool "STB04 DMA controller support (TD)"
+	depends on PPC4xx_DMA
+	default y
+
 config PPC_GEN550
 	bool
 	depends on 4xx
diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/platforms/4xx/Makefile linux-2.6.12/arch/ppc/platforms/4xx/Makefile
--- linux-2.6.12.orig/arch/ppc/platforms/4xx/Makefile	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/platforms/4xx/Makefile	2005-06-27 22:34:35.000000000 +0200
@@ -9,14 +9,14 @@
 obj-$(CONFIG_LUAN)		+= luan.o
 obj-$(CONFIG_OAK)		+= oak.o
 obj-$(CONFIG_OCOTEA)		+= ocotea.o
-obj-$(CONFIG_REDWOOD_5)		+= redwood5.o
+obj-$(CONFIG_IBM_PALLAS)	+= redwood5.o
 obj-$(CONFIG_REDWOOD_6)		+= redwood6.o
 obj-$(CONFIG_SYCAMORE)		+= sycamore.o
 obj-$(CONFIG_WALNUT)		+= walnut.o
 obj-$(CONFIG_XILINX_ML300)	+= xilinx_ml300.o

 obj-$(CONFIG_405GP)		+= ibm405gp.o
-obj-$(CONFIG_REDWOOD_5)		+= ibmstb4.o
+obj-$(CONFIG_IBM_PALLAS)	+= ibmstb4.o
 obj-$(CONFIG_NP405H)		+= ibmnp405h.o
 obj-$(CONFIG_REDWOOD_6)		+= ibmstbx25.o
 obj-$(CONFIG_440GP)		+= ibm440gp.o
diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/platforms/4xx/ibmstb4.c linux-2.6.12/arch/ppc/platforms/4xx/ibmstb4.c
--- linux-2.6.12.orig/arch/ppc/platforms/4xx/ibmstb4.c	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/platforms/4xx/ibmstb4.c	2005-06-26 21:48:33.000000000 +0200
@@ -11,6 +11,7 @@

 #include <linux/init.h>
 #include <asm/ocp.h>
+#include <asm/usb.h>
 #include <platforms/4xx/ibmstb4.h>

 static struct ocp_func_iic_data ibmstb4_iic0_def = {
@@ -22,7 +23,7 @@
 };
 OCP_SYSFS_IIC_DATA()

-struct ocp_def core_ocp[] __initdata = {
+struct ocp_def core_ocp[] = {
 	{ .vendor	= OCP_VENDOR_IBM,
 	  .function	= OCP_FUNC_16550,
 	  .index	= 0,
@@ -46,6 +47,7 @@
 	},
 	{ .vendor	= OCP_VENDOR_IBM,
 	  .function	= OCP_FUNC_IIC,
+	  .index    = 0,
 	  .paddr	= IIC0_BASE,
 	  .irq		= IIC0_IRQ,
 	  .pm		= IBM_CPM_IIC0,
@@ -54,6 +56,7 @@
 	},
 	{ .vendor	= OCP_VENDOR_IBM,
 	  .function	= OCP_FUNC_IIC,
+	  .index    = 1,
 	  .paddr	= IIC1_BASE,
 	  .irq		= IIC1_IRQ,
 	  .pm		= IBM_CPM_IIC1,
@@ -79,5 +82,76 @@
 	  .pm		= OCP_CPM_NA,
 	},
 	{ .vendor	= OCP_VENDOR_INVALID,
+	},
+};
+
+
+
+
+// Power up the USB subsection
+static int enable_usb(struct platform_device *pdev) {
+	u32 mask;
+
+	mask = 1 << (31 - USB0_IRQ);
+	mtdcr(DCRN_UIC_PR(UIC0), mfdcr(DCRN_UIC_PR(UIC0)) | mask);
+	mtdcr(DCRN_UIC_TR(UIC0), mfdcr(DCRN_UIC_TR(UIC0)) & ~mask);
+	return(0);
 	}
+
+// Power down the USB subsection
+static void disable_usb(struct platform_device *pdev) {
+	u32 mask;
+
+	mask = 1 << (31 - USB0_IRQ);
+	mtdcr(DCRN_UIC_PR(UIC0), mfdcr(DCRN_UIC_PR(UIC0)) & ~mask);
+	mtdcr(DCRN_UIC_TR(UIC0), mfdcr(DCRN_UIC_TR(UIC0)) | mask);
+}
+
+static struct usb_hcd_platform_data pd = {
+	.start = enable_usb,
+	.stop = disable_usb,
+};
+
+
+static struct resource ohci_usb_resources[] = {
+	[0] = {
+		.start  = USB0_BASE,
+		.end    = USB0_BASE + USB0_SIZE - 1,
+		.flags  = IORESOURCE_MEM,
+	},
+		[1] = {
+		.start  = USB0_IRQ,
+		.end    = USB0_IRQ,
+		.flags  = IORESOURCE_IRQ,
+	},
+};
+
+static u64 dma_mask = 0xffffffffULL;
+
+
+
+static struct platform_device ohci_usb_device = {
+	.name       = "ppc-soc-ohci",
+	.id     = 0,
+	.num_resources  = ARRAY_SIZE(ohci_usb_resources),
+					  .resource   = ohci_usb_resources,
+	.dev        = {
+		.dma_mask = &dma_mask,
+		.coherent_dma_mask = 0xffffffffULL,
+		.platform_data = &pd,
+	},
+};
+
+
+static struct platform_device *ibmstb4_devs[] __initdata = {
+	&ohci_usb_device,
 };
+
+
+//--------------------------------------------------------------------------
+static int __init ibmstb4_platform_add_devices(void) {
+	return(platform_add_devices(ibmstb4_devs, ARRAY_SIZE(ibmstb4_devs)));
+}
+
+arch_initcall(ibmstb4_platform_add_devices);
+
diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/platforms/4xx/ibmstb4.h linux-2.6.12/arch/ppc/platforms/4xx/ibmstb4.h
--- linux-2.6.12.orig/arch/ppc/platforms/4xx/ibmstb4.h	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/platforms/4xx/ibmstb4.h	2005-06-27 11:29:35.000000000 +0200
@@ -73,9 +73,9 @@
 #define OPB0_BASE	0x40000000
 #define GPIO0_BASE	0x40060000

+#define USB0_BASE	0x40010000
+#define USB0_SIZE	0xA0
 #define USB0_IRQ	18
-#define USB0_BASE	STB04xxx_MAP_IO_ADDR(0x40010000)
-#define USB0_EXTENT 4096

 #define IIC_NUMS 2
 #define UART_NUMS	3
@@ -160,20 +160,21 @@
 #define IBM_CPM_SC1	0x00008000	/* Smart Card 1 */
 #define IBM_CPM_USBSDRA	0x00004000	/* SDRAM 0 memory controller */
 #define IBM_CPM_XPT0	0x00002000	/* Transport - 54 Mhz */
-#define IBM_CPM_CBS	0x00001000	/* Cross Bar Switch */
+#define IBM_CPM_DDI12	(1 << (31-19))	/* Descrambler 1, 2 */
 #define IBM_CPM_GPT	0x00000800	/* GPTPWM */
 #define IBM_CPM_GPIO0	0x00000400	/* General Purpose IO 0 */
 #define IBM_CPM_DENC	0x00000200	/* Digital video Encoder */
 #define IBM_CPM_TMRCLK	0x00000100	/* CPU timers */
 #define IBM_CPM_XPT27	0x00000080	/* Transport - 27 Mhz */
 #define IBM_CPM_UIC	0x00000040	/* Universal Interrupt Controller */
-#define IBM_CPM_SSP	0x00000010	/* Modem Serial Interface (SSP) */
+#define IBM_CPM_GFX	(1 << (31-26))	/* Graphics Control */
+#define IBM_CPM_SSP	(1 << (31-27))	/* Modem Serial Interface (SSP) */
 #define IBM_CPM_UART2	0x00000008	/* Serial Control Port */
 #define IBM_CPM_DDIO	0x00000004	/* Descrambler */
 #define IBM_CPM_VID2	0x00000002	/* Video Decoder clock domain 2 */

 #define DFLT_IBM4xx_PM	~(IBM_CPM_CPU | IBM_CPM_EBIU | IBM_CPM_SDRAM1 \
-			| IBM_CPM_DMA | IBM_CPM_DMA1 | IBM_CPM_CBS \
+			| IBM_CPM_DMA | IBM_CPM_DMA1 | IBM_CPM_DDI12 \
 			| IBM_CPM_USBSDRA | IBM_CPM_XPT0 | IBM_CPM_TMRCLK \
 			| IBM_CPM_XPT27 | IBM_CPM_UIC )

diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/platforms/4xx/redwood5.c linux-2.6.12/arch/ppc/platforms/4xx/redwood5.c
--- linux-2.6.12.orig/arch/ppc/platforms/4xx/redwood5.c	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/platforms/4xx/redwood5.c	2005-06-27 23:19:21.000000000 +0200
@@ -18,6 +18,8 @@
 #include <linux/ioport.h>
 #include <asm/io.h>
 #include <asm/machdep.h>
+#include <linux/mtd/physmap.h>
+#include <linux/mtd/map.h>

 static struct resource smc91x_resources[] = {
 	[0] = {
@@ -49,9 +53,14 @@
 	return platform_add_devices(redwood5_devs, ARRAY_SIZE(redwood5_devs));
 }

-void __init
-redwood5_setup_arch(void)
-{
+
+#define CONFIG_DEBUG_BRINGUP 1
+
+void __init redwood5_setup_arch(void) {
+#if defined(CONFIG_DEBUG_BRINGUP) || defined(CONFIG_TRIPLEDRAGON)
+	const bd_t * const bip = &__res;
+#endif
+
 	ppc4xx_setup_arch();

 #ifdef CONFIG_DEBUG_BRINGUP
@@ -55,48 +64,58 @@
 	ppc4xx_setup_arch();

 #ifdef CONFIG_DEBUG_BRINGUP
+
 	printk("\n");
-	printk("machine\t: %s\n", PPC4xx_MACHINE_NAME);
+	printk("machine: %s\n", PPC4xx_MACHINE_NAME);
+	printk("bi_s_version (bd_t struct's version)  : %s\n", bip->bi_s_version);
+	printk("bi_r_version (bios/bootloader version): %s\n", bip->bi_r_version);
 	printk("\n");
-	printk("bi_s_version\t %s\n",      bip->bi_s_version);
-	printk("bi_r_version\t %s\n",      bip->bi_r_version);
-	printk("bi_memsize\t 0x%8.8x\t %dMBytes\n", bip->bi_memsize,bip->bi_memsize/(1024*1000));
-	printk("bi_enetaddr %d\t %2.2x%2.2x%2.2x-%2.2x%2.2x%2.2x\n", 0,
-	bip->bi_enetaddr[0], bip->bi_enetaddr[1],
-	bip->bi_enetaddr[2], bip->bi_enetaddr[3],
-	bip->bi_enetaddr[4], bip->bi_enetaddr[5]);
+	printk("bi_memsize : 0x%.8x (%.2d)MBytes\n", bip->bi_memsize, bip->bi_memsize/(1024*1000));
+	printk("bi_memsize1: 0x%.8x (%.2d)MBytes\n", bip->bi_memsize1, bip->bi_memsize1/(1024*1000));
+	printk("bi_enetaddr: %.2x:%.2x:%.2x:%.2x:%.2x:%.2x\n",
+		   bip->bi_enetaddr[0], bip->bi_enetaddr[1], bip->bi_enetaddr[2],
+		   bip->bi_enetaddr[3], bip->bi_enetaddr[4], bip->bi_enetaddr[5]);

-	printk("bi_intfreq\t 0x%8.8x\t clock:\t %dMhz\n",
+	printk("bi_intfreq (CPU speed) : 0x%.8x (%d)Mhz\n",
 	       bip->bi_intfreq, bip->bi_intfreq/ 1000000);

-	printk("bi_busfreq\t 0x%8.8x\t plb bus clock:\t %dMHz\n",
+	printk("bi_busfreq (PLB clock) : 0x%.8x (%d)MHz\n",
 		bip->bi_busfreq, bip->bi_busfreq / 1000000 );
-	printk("bi_tbfreq\t 0x%8.8x\t TB freq:\t %dMHz\n",
+	printk("bi_tbfreq (SW timebase): 0x%.8x (%d)MHz\n",
 	       bip->bi_tbfreq, bip->bi_tbfreq/1000000);

+	printk("bi_opb_busfreq         : 0x%.8x (%d)MHz\n",
+		   bip->bi_opb_busfreq, bip->bi_opb_busfreq / 1000000);
+	printk("bi_iic_fast enable     : [0] %.1x, [1] %.1x\n",
+		   bip->bi_iic_fast[0], bip->bi_iic_fast[1]);
+	printk("flash_base             : 0x%.8x\n", bip->flash_base);
+
 	printk("\n");
+
 #endif
+	{
 	device_initcall(redwood5_platform_add_devices);
 }

-void __init
-redwood5_map_io(void)
-{
-	int i;
+#if defined(CONFIG_TRIPLEDRAGON)
+	physmap_configure(bip->flash_base, -bip->flash_base, 2, NULL);
+#endif
+
+	/* Identify the system */
+	printk(KERN_INFO "MNC Ltd. DBx3000\n");
+	printk(KERN_INFO "http://www.mncltd.com\n");
+}
+
+void __init redwood5_map_io(void) {
+	unsigned int i, v, p;

 	ppc4xx_map_io();
 	for (i = 0; i < 16; i++) {
-	 unsigned long v, p;
-
 	/* 0x400x0000 -> 0xe00x0000 */
 	p = 0x40000000 | (i << 16);
 	v = STB04xxx_IO_BASE | (i << 16);
-
-	io_block_mapping(v, p, PAGE_SIZE,
-		 _PAGE_NO_CACHE | pgprot_val(PAGE_KERNEL) | _PAGE_GUARDED);
+		io_block_mapping(v, p, PAGE_SIZE, _PAGE_NO_CACHE | pgprot_val(PAGE_KERNEL) | _PAGE_GUARDED);
 	}
-
-
 }

 void __init
diff -Xxfiles -rNBwdu linux-2.6.12.orig/arch/ppc/platforms/4xx/redwood5.h linux-2.6.12/arch/ppc/platforms/4xx/redwood5.h
--- linux-2.6.12.orig/arch/ppc/platforms/4xx/redwood5.h	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/arch/ppc/platforms/4xx/redwood5.h	2005-06-27 23:18:33.000000000 +0200
@@ -21,6 +21,23 @@

 #ifndef __ASSEMBLY__
 typedef struct board_info {
+
+#if defined(CONFIG_TRIPLEDRAGON)
+
+	unsigned char   bi_s_version[4];	/* Version of this structure */
+	unsigned char   bi_r_version[30];	/* Version of the IBM ROM */
+	unsigned int    bi_memsize;		/* DRAM installed, in bytes */
+	unsigned int    bi_memsize1;		/* field shouldn't exist */
+	unsigned char   bi_enetaddr[6];		/* Ethernet MAC address */
+	unsigned int    bi_intfreq;		/* Processor speed, in Hz */
+	unsigned int    bi_busfreq;		/* Bus speed, in Hz */
+	unsigned int    bi_tbfreq;		/* Software timebase freq */
+	unsigned int    bi_opb_busfreq;		/* OPB Bus speed, in Hz */
+	int     bi_iic_fast[2];		/* Use fast i2c mode */
+	unsigned char   *bi_cmdline;		/* bios kernel commandline */
+	unsigned int    flash_base;		/* flash base address */
+
+#else
 	unsigned char	bi_s_version[4];	/* Version of this structure */
 	unsigned char	bi_r_version[30];	/* Version of the IBM ROM */
 	unsigned int	bi_memsize;		/* DRAM installed, in bytes */
@@ -29,7 +46,10 @@
 	unsigned int	bi_intfreq;		/* Processor speed, in Hz */
 	unsigned int	bi_busfreq;		/* Bus speed, in Hz */
 	unsigned int	bi_tbfreq;		/* Software timebase freq */
+#endif
+
 } bd_t;
+
 #endif /* !__ASSEMBLY__ */


@@ -48,7 +68,11 @@

 #define BASE_BAUD		(378000000 / 18 / 16)

+#if defined(CONFIG_TRIPLEDRAGON)
+#define PPC4xx_MACHINE_NAME	"TRIPLEDRAGON STB"
+#else
 #define PPC4xx_MACHINE_NAME	"IBM Redwood5"
+#endif

 #endif /* __ASM_REDWOOD5_H__ */
 #endif /* __KERNEL__ */
diff -Xxfiles -rNBwdu linux-2.6.12.orig/drivers/net/Kconfig linux-2.6.12/drivers/net/Kconfig
--- linux-2.6.12.orig/drivers/net/Kconfig	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/drivers/net/Kconfig	2005-06-25 22:22:58.000000000 +0200
@@ -794,7 +794,7 @@
 	tristate "SMC 91C9x/91C1xxx support"
 	select CRC32
 	select MII
-	depends on NET_ETHERNET && (ARM || REDWOOD_5 || REDWOOD_6 || M32R || SUPERH)
+	depends on NET_ETHERNET && (ARM || REDWOOD_5 || REDWOOD_6 || M32R || SUPERH || TRIPLEDRAGON)
 	help
 	  This is a driver for SMC's 91x series of Ethernet chipsets,
 	  including the SMC91C94 and the SMC91C111. Say Y if you want it
diff -Xxfiles -rNBwdu linux-2.6.12.orig/drivers/net/smc91x.h linux-2.6.12/drivers/net/smc91x.h
--- linux-2.6.12.orig/drivers/net/smc91x.h	2005-06-17 21:48:29.000000000 +0200
+++ linux-2.6.12/drivers/net/smc91x.h	2005-06-27 22:48:05.000000000 +0200
@@ -55,7 +55,7 @@
 #define SMC_insw(a, r, p, l)	readsw((a) + (r), p, l)
 #define SMC_outsw(a, r, p, l)	writesw((a) + (r), p, l)

-#elif defined(CONFIG_REDWOOD_5) || defined(CONFIG_REDWOOD_6)
+#elif defined(CONFIG_IBM_PALLAS) || defined(CONFIG_REDWOOD_6)

 /* We can only do 16-bit reads and writes in the static memory space. */
 #define SMC_CAN_USE_8BIT	0
